File: src/classes/extra_frame.py
from .pin import Pin
from .base_frame import BaseFrame

class ExtraFrame(BaseFrame):
  def __init__(self, number: int):
    super().__init__(self, number)
    self.bonus_points: int = 0
    self.knock_down_pins: dict[str, list[Pin]] = {
      'first_throw': [],
      'second_throw': [],
      'third_throw': [],
    }
    self.number: int = number
    self.standing_pins: list[Pin] = []
    self.total_score: int = 0

  @property
  def round_score(self) -> int:
    return len(self.knock_down_pins['first_throw']) + len(self.knock_down_pins['second_throw']) + len(self.knock_down_pins['third_throw'])

  @property
  def spares(self) -> dict[str, list[bool]]:
    return {
      'first_pair_throw': len(self.knock_down_pins['first_throw']) + len(self.knock_down_pins['second_throw']) == 10,
      'second_pair_throw': len(self.knock_down_pins['second_throw']) + len(self.knock_down_pins['third_throw']) == 10,
    }

  @property
  def strikes(self) -> dict[str, list[bool]]:
    return {
      'first_throw': len(self.knock_down_pins['first_throw']) == 10,
      'second_throw': len(self.knock_down_pins['second_throw']) == 10,
      'third_throw': len(self.knock_down_pins['third_throw']) == 10,
    }

  @property
  def third_throw_score(self) -> int:
    return len(self.knock_down_pins['third_throw'])

File: Bowling/src/classes/ExtraFrame.cs
namespace Bowling;

public class ExtraFrame : BaseFrame
{
  public ExtraFrame(int Number) : base(Number)
  {
    BonusPoints = 0;
    KnockedDownPins = new()
    {
        { "FirstThrow", new() },
        { "SecondThrow", new() },
        { "ThirdThrow", new() },
    };
    this.Number = Number;
    StandingPins = new();
    TotalScore = 0;
  }

  public override int RoundScore
  {
    get => KnockedDownPins["FirstThrow"].Count + KnockedDownPins["SecondThrow"].Count + KnockedDownPins["ThirdThrow"].Count;
  }
  public Dictionary<string, bool> Spares
  {
    get => new()
    {
        { "FirstPairThrow", KnockedDownPins["FirstThrow"].Count + KnockedDownPins["SecondThrow"].Count == 10 },
        { "SecondPairThrow", KnockedDownPins["SecondThrow"].Count + KnockedDownPins["ThirdThrowThrow"].Count == 10 }
    };
  }
  public Dictionary<string, bool> Strikes
  {
    get => new()
  {
      { "FirstThrow", KnockedDownPins["FirstThrow"].Count == 10 },
      { "SecondThrow", KnockedDownPins["SecondThrow"].Count == 10 },
      { "ThirdThrow", KnockedDownPins["ThirdThrow"].Count == 10 },
  };
  }
  public int ThirdThrowScore
  {
    get => KnockedDownPins["ThirdThrow"].Count;
  }
}
